apis/database/v1alpha3/firewallrule.go

```diff
@@ -0,0 +1,153 @@
+
+	// FirewallRuleProperties - Resource properties.
+	FirewallRuleProperties `json:"properties"`
+}
+
+// +kubebuilder:object:root=true
+
+// A PostgreSQLServerFirewallRule is a managed resource that represents
+// an Azure PostgreSQL Database virtual network rule.
+// +kubebuilder:printcolumn:name="STATUS",type="string",JSONPath=".status.bindingPhase"
+// +kubebuilder:printcolumn:name="STATE",type="string",JSONPath=".status.state"
+// +kubebuilder:printcolumn:name="AGE",type="date",JSONPath=".metadata.creationTimestamp"
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:scope=Cluster,categories={crossplane,managed,azure}
+type PostgreSQLServerFirewallRule struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   PostgreSQLFirewallRuleSpec `json:"spec"`
+	Status FirewallRuleStatus         `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// PostgreSQLServerFirewallRuleList contains a list of PostgreSQLServerFirewallRule.
+type PostgreSQLServerFirewallRuleList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []PostgreSQLServerFirewallRule `json:"items"`
+}
+
+// A MySQLFirewallRuleSpec defines the desired state of a MySQLFirewallRule.
+type MySQLFirewallRuleSpec struct {
+	runtimev1alpha1.ResourceSpec `json:",inline"`
+
+	// ServerName - Name of the Virtual Network Rule's server.
+	ServerName string `json:"serverName,omitempty"`
+
+	// ServerNameRef - A reference to the Virtual Network Rule's MySQLServer.
+	ServerNameRef *runtimev1alpha1.Reference `json:"serverNameRef,omitempty"`
+
+	// ServerNameSelector - Selects a MySQLServer to reference.
+	ServerNameSelector *runtimev1alpha1.Selector `json:"serverNameSelector,omitempty"`
+
+	// ResourceGroupName - Name of the Virtual Network Rule's resource group.
+	ResourceGroupName string `json:"resourceGroupName,omitempty"`
+
+	// ResourceGroupNameRef - A reference to a ResourceGroup object to retrieve
+	// its name
+	ResourceGroupNameRef *runtimev1alpha1.Reference `json:"resourceGroupNameRef,omitempty"`
+
+	// ResourceGroupNameSelector - Selects a ResourceGroup to reference.
+	ResourceGroupNameSelector *runtimev1alpha1.Selector `json:"resourceGroupNameSelector,omitempty"`
+
+	// FirewallRuleProperties - Resource properties.
+	FirewallRuleProperties `json:"properties"`
+}
+
+// +kubebuilder:object:root=true
+
+// A MySQLServerFirewallRule is a managed resource that represents an
+// Azure MySQL Database virtual network rule.
+// +kubebuilder:printcolumn:name="STATUS",type="string",JSONPath=".status.bindingPhase"
+// +kubebuilder:printcolumn:name="STATE",type="string",JSONPath=".status.state"
+// +kubebuilder:printcolumn:name="AGE",type="date",JSONPath=".metadata.creationTimestamp"
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:scope=Cluster,categories={crossplane,managed,azure}
+type MySQLServerFirewallRule struct {
```

Comment:
Is this rule generic or specific per DB?

Reply:
I think it's specific per DB. The types are identical but the Azure API seems to have different endpoints depending on the type of database you're using, for example:

https://docs.microsoft.com/en-us/rest/api/postgresql/firewallrules/createorupdate uses `PUT https://management.azure.com/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DBforPostgreSQL/servers/{serverName}/firewallRules/{firewallRuleName}?api-version=2017-12-01`

https://docs.microsoft.com/en-us/rest/api/mysql/firewallrules/createorupdate uses `PUT https://management.azure.com/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DBforMySQL/servers/{serverName}/firewallRules/{firewallRuleName}?api-version=2017-12-01`

Note `Microsoft.DBforPostgreSQL` in the first URL vs `Microsoft.DBforMySQL` in the second.

We've historically modelled this as two different types in Crossplane - e.g. `MySQLServer` vs `PostgreSQLServer`. This is a little different from other clouds, where the SQL engine variant is typically a spec field like `spec.engine: mysql`.